internal/kubernetes/drainSchedule.go

```diff
@@ -0,0 +1,177 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+	core "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+)
+
+const (
+	SetConditionTimeout     = 10 * time.Second
+	SetConditionRetryPeriod = 50 * time.Millisecond
+)
+
+type DrainScheduler interface {
+	HasSchedule(name string) (has, running, failed bool)
+	Schedule(node *v1.Node) (time.Time, error)
+	DeleteSchedule(name string)
+}
+
+type DrainSchedules struct {
+	sync.Mutex
+	schedules map[string]*schedule
+
+	lastDrainScheduledFor time.Time
+	period                time.Duration
+
+	logger        *zap.Logger
+	drainer       Drainer
+	eventRecorder record.EventRecorder
+}
+
+func NewDrainSchedules(drainer Drainer, eventRecorder record.EventRecorder, period time.Duration, logger *zap.Logger) DrainScheduler {
+	return &DrainSchedules{
+		schedules: map[string]*schedule{},
+		//lastDrainScheduledFor: time.Now(),
+		period:        period,
+		logger:        logger,
+		drainer:       drainer,
+		eventRecorder: eventRecorder,
+	}
+}
+
+func (d *DrainSchedules) HasSchedule(name string) (has, running, failed bool) {
+	d.Lock()
+	defer d.Unlock()
+	sched, ok := d.schedules[name]
+	if !ok {
+		return false, false, false
+	}
+	return true, sched.running, sched.failed
+}
+
+func (d *DrainSchedules) DeleteSchedule(name string) {
+	d.Lock()
+	defer d.Unlock()
+	if s, ok := d.schedules[name]; ok {
+		s.timer.Stop()
+	} else {
+		d.logger.Error("Failed schedule deletion", zap.String("key", name))
+	}
+	delete(d.schedules, name)
+}
+
+func (d *DrainSchedules) WhenNextSchedule() time.Time {
+	// compute drain schedule time
+	sooner := time.Now().Add(SetConditionTimeout + time.Second)
+	when := d.lastDrainScheduledFor.Add(d.period)
+	if when.Before(sooner) {
+		when = sooner
+	}
+	return when
+}
+
+func (d *DrainSchedules) Schedule(node *v1.Node) (time.Time, error) {
+	d.Lock()
+	if sched, ok := d.schedules[node.GetName()]; ok {
+		d.Unlock()
+		return sched.when, NewAlreadyScheduledError() // we already have a schedule planned
+	}
+
+	// compute drain schedule time
+	when := d.WhenNextSchedule()
+	d.lastDrainScheduledFor = when
+	d.schedules[node.GetName()] = d.newSchedule(node, when)
+	d.Unlock()
+
+	// Mark the node with the condition stating that drain is scheduled
+	if err := RetryWithTimeout(
+		func() error {
+			return d.drainer.MarkDrain(node, when, time.Time{}, false)
+		},
+		SetConditionRetryPeriod,
+		SetConditionTimeout,
+	); err != nil {
+		// if we cannot mark the node, let's remove the schedule
+		d.DeleteSchedule(node.GetName())
+		return time.Time{}, err
+	}
+	return when, nil
+}
+
+type schedule struct {
+	when    time.Time
+	running bool
+	failed  bool
+	finish  time.Time
+	timer   *time.Timer
+}
+
+func (d *DrainSchedules) newSchedule(node *v1.Node, when time.Time) *schedule {
```

Comment:
In reading the body of the `time.AfterFunc` I see that we're issuing a `Drain` - it looks like that drain occurs before we've marked the node for draining. I'm jumping back to the caller to notice that we call `MarkDrain` immediately after calling `newSchedule`.

I asked myself why not put the the contents of line 97 inside of `d.newSchedule` to encapsulate all of the machinery around:

- RetryWithTimeout(Marking a node for draining).
- Create the schedule timer only if the drain succeeds.

Therefore avoiding the `DeleteSchedule` if the `RetryWithTimeout(Marking a node for draining)` bit fails - we just don't create the schedule if we fail to mark the node.

It looks like we're returning from `newSchedule` quickly in order to limit the duration we've acquired the lock for?

Reply:
The Drain can't in theory occur before the MarkDrain because the schedule is always set with an extended delay of `SetConditionTimeout` in function `WhenNextSchedule`.
If we place the `RetryWithTimeout` before the call to `newSchedule` we are going to hold the lock over the `RetryWithTimeout` which I wanted to avoid, because holding the lock for 10 sec is not really good if we have multiple workers.